samples/algorithms/sudoku-grover/Sudoku.qs

```diff
@@ -0,0 +1,203 @@
+    ///         empty square #0 can not have the same color/number as empty call #1 
+    ///         empty square #1 and #2 can not have the same color/number (same column)
+    ///    startingNumberConstraints = [(0, 2),(0, 1),(0, 3),(1, 1),(1, 2),(1, 0),(2, 1),(2, 2),(2, 3)]
+    ///         empty square #0 can not have values 2,1,3 because same row/column/2x2grid
+    ///         empty square #1 can not have values 1,2,0 because same row/column/2x2grid
+    ///    Results = [0,3,0] i.e. Empty Square #0 = 0, Empty Square #1 = 3, Empty Square #2 = 0
+    operation SolvePuzzle(numVertices : Int, size : Int, emptySquareEdges : (Int, Int)[], 
+        startingNumberConstraints: (Int, Int)[]) : (Bool, Int[]) {
+        // for size = 4x4 grid
+        mutable bitsPerColor = 2; 
+        mutable oracle = VertexColoringOracle(numVertices, bitsPerColor, emptySquareEdges, startingNumberConstraints, _, _);
+        if (size == 9) {
+            // if size == 9x9 grid, the solution has 9 colors so we need 
+            // 4 bits per color
+            set bitsPerColor = 4;
+            // Although we could use VertexColoringOracle for 9x9, we would
+            // have to add restrictions on each color to not allow colors 8 
+            // thru 15. This could be achieved by adding these to 
+            // startNumberConstraints. However, this did not scale well with 
+            // the simulator, and so instead we use 
+            // VertexColoringOracle4Bit9Color which has the 9 color 
+            // restriction built in.
+            set oracle = VertexColoringOracle4Bit9Color(numVertices, emptySquareEdges, startingNumberConstraints, _, _);
+        } elif (size != 4) {
+            fail $"Cannot set size {size}: only a grid size of 4x4 or 9x9 is supported";
+        }
+        let numIterations = NIterations(bitsPerColor * numVertices);
+        Message($"Running Quantum test with #Vertex = {numVertices}");
+        Message($"   Bits Per Color = {bitsPerColor}");
+        Message($"   emptySquareEdges = {emptySquareEdges}");
+        Message($"   startingNumberConstraints = {startingNumberConstraints}");
+        Message($"   estimated #iterations needed = {numIterations}");
+        Message($"   size of Sudoku grid = {size}x{size}");
+        let coloring = FindColorsWithGrover(numVertices, bitsPerColor, numIterations, oracle);
+
+        Message($"Got sudoku solution: {coloring}");
+        if (IsSudokuSolutionValid(size, emptySquareEdges, startingNumberConstraints, coloring)) {
+           Message($"Got valid sudoku solution: {coloring}");
+           return (true, coloring);
+        } else {
+           Message($"Got invalid sudoku solution: {coloring}");
+           return (false, coloring);
+        }
+    }
+
+    /// # Summary
+    /// Estimate the number of interations required for solution
+    ///
+    /// # Input
+    /// ## nQubits
+    /// The number of qubits being used
+    function NIterations(nQubits : Int) : Int {
```

Comment:
This is correct for an amplitude amplification problem with a single correct solution, but may need adapted when there are multiple solutions.